[InlineCost] Cache collectEphemeralValues() to save compile time

vporpo opened the pull request:
`CallAnalyzer::analyze()` can take several hours to run if the function contains thousands of @llvm.assume() calls and there are thousands of callsites. The time is spent in `collectEphemeralvalues()`.
This patch adds caching and will only collect the ephemeral values once per function.

Commit:
[InlineCost] Use EphemeralValuesCache to improve compile time

`CallAnalyzer::analyze()` can take several hours to run if the function contains
thousands of @llvm.assume() calls and there are thousands of callsites.
The time is spent in `collectEphemeralvalues()`.
This patch adds caching to InlineCost and will only collect the ephemeral values
once per function.

aeubanks commented:
can we make this a proper pass manager analysis and query the analysis manager instead of an inliner-specific cache, just like `GetAssumptionCache`? that way any pass that wants to calculate ephemeral values can do so the same way and potentially reuse it. I believe if we go down this route, we need to invalidate analyses on the callee a bit more aggressively, as in moving [this](https://github.com/llvm/llvm-project/blob/e4a6e2eb7176e677eb9a9c45913a284d374d60c5/llvm/lib/Transforms/IPO/Inliner.cpp#L529) to [here](https://github.com/llvm/llvm-project/blob/e4a6e2eb7176e677eb9a9c45913a284d374d60c5/llvm/lib/Transforms/IPO/Inliner.cpp#L392), in the case we are inlining a recursive call (we probably should already be doing this anyway).

Commit:
[Analysis][EphemeralValuesCache] Adds an ephemeral values cache analysis pass

This analysis pass collects the ephemeral values of a function and caches them
for fast lookups. The collection of the ephemeral values is done lazily when
the user calls `EphemeralValuesCache::ephValues()`.

vporpo commented:
@aeubanks perhaps something along these lines ?

nikic commented on llvm/include/llvm/Analysis/EphemeralValuesCache.h:
```diff
@@ -0,0 +1,61 @@
+namespace llvm {
+
+class Function;
+template <class T> class SmallPtrSetImpl;
```
Unnecessary forward declare, you include the header.

vporpo replied:
Done

nikic commented on llvm/include/llvm/Analysis/EphemeralValuesCache.h:
```diff
@@ -0,0 +1,61 @@
+    EphValues.clear();
+    Collected = false;
+  }
+  const SmallPtrSet<const Value *, 32> &ephValues() {
```
```suggestion
  const SmallPtrSetImpl<const Value *> &ephValues() {
```

vporpo replied:
Done

nikic commented on llvm/include/llvm/Analysis/EphemeralValuesCache.h:
```diff
@@ -0,0 +1,61 @@
+
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/Analysis/AssumptionCache.h"
+#include "llvm/Analysis/CodeMetrics.h"
```
I don't think we need CodeMetrics in the header?

vporpo replied:
Done

nikic commented on llvm/lib/Transforms/IPO/Inliner.cpp:
```diff
@@ -388,6 +389,8 @@ PreservedAnalyses InlinerPass::run(LazyCallGraph::SCC &InitialC,
         continue;
       }
+      // The caller was modified, so invalidate Ephemeral Values.
+      FAM.getResult<EphemeralValuesAnalysis>(F).clear();
```
Hm, shouldn't we be invalidating all analyses on the function here, rather than just EphemeralValues?

vporpo replied:
I think so, but let's make this a separate patch in case it breaks something. I added a TODO.

nikic replied:
Okay, that sounds reasonable.

nikic reviewed (commented):
Looks reasonable to me.

vporpo commented:
@aeubanks any thoughts or comments ?

nikic commented on llvm/include/llvm/Analysis/EphemeralValuesCache.h:
```diff
@@ -0,0 +1,59 @@
+#define LLVM_ANALYSIS_EPHEMERALVALUESCACHE_H
+
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/Analysis/AssumptionCache.h"
```
I think you could forward-declare AssumptionCache?

vporpo replied:
Done

nikic commented on llvm/lib/Analysis/InlineCost.cpp:
```diff
@@ -2781,11 +2790,15 @@ InlineResult CallAnalyzer::analyze() {
+  // ephemeral values cache if available.
+  SmallPtrSet<const Value *, 32> EphValuesStorage;
+  const SmallPtrSetImpl<const Value *> *EphValues = &EphValuesStorage;
+  auto &AC = GetAssumptionCache(F);
```
Can move the GetAssumptionCache() into the branch that uses it.

vporpo replied:
Done

nikic commented on llvm/include/llvm/Analysis/InlineCost.h:
```diff
@@ -273,14 +274,14 @@ int getCallsiteCost(const TargetTransformInfo &TTI, const CallBase &Call,
+    function_ref<BlockFrequencyInfo &(Function &)> GetBFI = nullptr,
+    ProfileSummaryInfo *PSI = nullptr, OptimizationRemarkEmitter *ORE = nullptr,
+    std::optional<function_ref<EphemeralValuesCache &(Function &)>>
+        GetEphValuesCache = std::nullopt);
```
function_ref accepts nullptr, so I think it would be better to use `function_ref<EphemeralValuesCache &(Function &)> GetEphValuesCache = nullptr` here, to save the std::optional wrapper. See GetBFI above.

vporpo replied:
Done

nikic reviewed (approved):
LGTM

aeubanks commented on llvm/include/llvm/Analysis/EphemeralValuesCache.h:
```diff
@@ -0,0 +1,60 @@
+    EphValues.clear();
+    Collected = false;
+  }
+  const SmallPtrSetImpl<const Value *> &ephValues() {
```
why do we need an extra layer of indirection? can't we just eagerly calculate the ephemeral values in `run()`?

vporpo replied:
Do you mean removing the indirection by incorporating the functionality of the `EphemeralValuesCache` inside the `EphemeralValuesAnalysis` class? Isn't it better to follow the two-class design where you implement the analysis in one class and have a separate wrapper class for the pass in case someone needs to use the analysis without a pass ?

As for the eager vs lazy approach, my guess is that lazy is preferable because it may save you some compile time if you don't end up calling the analysis, but I don't feel strongly about it.

aeubanks replied:
`CodeMetrics::collectEphemeralValues()` is the API if someone wants to calculate ephemeral values, I don't think we need another layer on top of that. I think `EphemeralValuesAnalysis::Result` could just be `SmallPtrSet<const Value *, 32>`

the pass should only call `FAM.getAnalysis` if it's going to use it, so the laziness should be done at the caller level

aeubanks reviewed (commented):
sorry, I wasn't getting email updates for this PR for some reason
